ebpf/kernel/llb_kern_ct.c

```diff
@@ -5,6 +5,63 @@
  * SPDX-License-Identifier: GPL-2.0
  */
 
+#define CT_CTR_SID      (0)
+#define CT_CTR_MAX_SID (250000)
+
+struct dp_ct_ctrtact {
+  struct dp_cmn_act ca; /* Possible actions :
+                         * None (just place holder) 
+                         */
+  struct bpf_spin_lock lock;
+  __u32 counter; 
+};
+
+#ifdef HAVE_LEGACY_BPF_MAPS
+
+struct bpf_map_def SEC("maps") ct_ctr = {
+  .type = BPF_MAP_TYPE_ARRAY,
+  .key_size = sizeof(__u32),
+  .value_size = sizeof(dp_ct_ctrtact),
+  .max_entries = 1 
+};
+
+#else
+
+struct {
+  __uint(type,        BPF_MAP_TYPE_ARRAY);
+  __type(key,         __u32);
+  __type(value,       struct dp_ct_ctrtact);
+  __uint(max_entries, 1);
+} ct_ctr SEC(".maps");
+
+#endif
+
+static __u32
+dp_ct_get_newctr(void)
+{
+  __u32 k = 0;
+  __u32 v = 0;
+  struct dp_ct_ctrtact *ctr;
+
+  ctr = bpf_map_lookup_elem(&ct_ctr, &k);
+
+  if (ctr == NULL) {
+    return 0;
+  }
+
+  /* FIXME - We can potentially do a percpu array and do away
+   *         with the locking here
+   */ 
+  bpf_spin_lock(&ctr->lock);
+  v = ++ctr->counter;
+  bpf_spin_unlock(&ctr->lock);
+
+  /* Essentially allocation starts from idx-2,4,8... */
+  v <<= 1;
+  v = (v + CT_CTR_SID) % CT_CTR_MAX_SID;
+  return v;
```

Comment:
Overall changes look good. There is one cause of concern. What happens when CT ctr indexes overflow ??  It can potentially cause to occupy used ct entries !!

Reply:
This is a  possible scenario and will be addressed in a future patch.